### PR TITLE
Implementing full width responsive non fast-fetch path

### PR DESCRIPTION
**Pull request by riklund**
This makes it possible for an ad network to specify an aspect ratio in the config, and if this is specified ad units with data-full-width set will then be expanded to 100% of the device width, aligned with the device edges and with she specified aspect ratio.

A default height needs to be provided for the ad unit in case changing the height to match the aspect ratio fails (for example if it's above the fold).

Tried to keep it reasonably close to the AdSense a4a path (https://github.com/ampproject/amphtml/pull/11231).

### CODE REVIEW

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js [hunk 1]
```diff
@@ -357,4 +359,170 @@ describes.realWin('amp-ad-3p-impl', {
...
+        container.appendChild(element);
+        containerContainer.appendChild(container);
+        doc.body.appendChild(containerContainer);
+
+        impl = new AmpAd3PImpl(element);
+        impl.element.style.display = 'block';
+        impl.element.style.position = 'relative';
+        impl.element.style.top = '150vh';
+
+          // Stub out vsync tasks to run immediately.
```
Comment (zhouyx) [inline extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: nit: indentation.

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js [hunk 2]
```diff
@@ -357,4 +359,170 @@ describes.realWin('amp-ad-3p-impl', {
...
+        impl.getVsync().run = (vsyncTaskSpec, vsyncState) => {
+          if (vsyncTaskSpec.measure) {
+            vsyncTaskSpec.measure(vsyncState);
+          }
+          if (vsyncTaskSpec.mutate) {
+            vsyncTaskSpec.mutate(vsyncState);
+          }
+        };
+
+          // Fix the viewport to a consistent size to that the test doesn't
```
Comment (zhouyx) [inline extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: nit: indentation

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js [hunk 3]
```diff
@@ -357,4 +359,170 @@ describes.realWin('amp-ad-3p-impl', {
...
+      let containerContainer, container;
+      let iframe;
+      let viewer;
+      let element;
+      let impl;
+
+      function buildImpl(config) {
+        containerContainer = doc.createElement('div');
+        container = doc.createElement('div');
+          // Create an element with horizontal margins for the ad to break out
```
Comment (zhouyx) [inline extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: nit: indentation

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js [hunk 4]
```diff
@@ -357,4 +359,170 @@ describes.realWin('amp-ad-3p-impl', {
...
+          // of.
+        containerContainer.style.marginLeft = '5px';
+        containerContainer.style.marginRight = '9px';
+
+          // Create an element with horizontal margins for the ad to break out
```
Comment (zhouyx) [inline extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: nit: indentation

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js [hunk 5]
```diff
@@ -357,4 +359,170 @@ describes.realWin('amp-ad-3p-impl', {
...
+    describe('#onLayoutMeasure', () => {
+        // Nested elements to contain the ad. (container contains the ad, and
+        // containerContainer contains that container.)
+      let containerContainer, container;
+      let iframe;
+      let viewer;
+      let element;
+      let impl;
+
+      function buildImpl(config) {
```
Comment (zhouyx) [inline extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: same here: can we reuse the `createAmpAd` function.

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js [hunk 6]
```diff
@@ -357,4 +359,170 @@ describes.realWin('amp-ad-3p-impl', {
...
+
+        config.type = '_ping_';
+
+        element = createElementWithAttributes(doc, 'amp-ad', config);
+        iframe = doc.createElement('iframe');
+
+        element.appendChild(iframe);
+        container.appendChild(element);
+        containerContainer.appendChild(container);
+        doc.body.appendChild(containerContainer);
```
Comment (zhouyx) [inline extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: use `win.doc.body` instead of `window.doc` so that it will be appended to the created env.win for single test. And there will be no need for you to remove containerContainer every time. 
Applies to all usage below.

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js [hunk 7]
```diff
@@ -357,4 +359,170 @@ describes.realWin('amp-ad-3p-impl', {
...
+      let iframe;
+      let viewer;
+      let element;
+      let impl;
+
+      function buildImpl(config) {
+        containerContainer = doc.createElement('div');
+        container = doc.createElement('div');
+          // Create an element with horizontal margins for the ad to break out
+          // of.
```
Comment (zhouyx) [inline extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: can we test against body padding as well?

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js [hunk 8]
```diff
@@ -144,6 +169,36 @@ export class AmpAd3PImpl extends AMP.BaseElement {
...
+    }
+  }
+
+  /**
+   * Sets the fullWidthState_ variable.
+   * @private
+   */
+  computeFullWidthState_() {
+    const hasFullWidth = this.element.hasAttribute('data-full-width');
+    const hasFullVW = this.element.getAttribute('width') == '100vw';
```
Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad-3p-impl.js]: throw use error? If width is not `100vw`

File: examples/a4a-fullwidth.amp.html [hunk 9]
```diff
@@ -81,7 +81,8 @@ <h1>Content!</h1>
           data-ad-client="ca-pub-2383777339857329"
           data-ad-slot="4121425836"
           data-adtest="on"
-          data-auto-format="rspv">
+          data-auto-format="rspv"
+          data-full-width>
```
[the remark this answers is not in the document]
Comment (riklund) [reply examples/a4a-fullwidth.amp.html]: It does if I disable the fast fetch cache manually, though.
I think the intent is to make the same ad slot work both in the a4a and 3p path?

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js [hunk 10]
```diff
@@ -33,12 +33,34 @@ import {user, dev} from '../../../src/log';
...
 /** @const {!string} Tag name for 3P AD implementation. */
 export const TAG_3P_IMPL = 'amp-ad-3p-impl';
 
+/**
+ * The full width state of this ad unit. The ad unit can request expansion to
+ * full-width by setting the width to 100vw and setting the attribute
+ * "data-full-width", if supported by the ad network.
+ * @enum {number}
+ */
+const FullWidthState = {
```
[the remark this answers is not in the document]
Comment (riklund) [reply extensions/amp-ad/0.1/amp-ad-3p-impl.js]: Why? With an enum we can ensure the state is always valid - with booleans we'd need to worry about keeping them in sync, for example it's not impossible to somehow get into the state with isFullWidthRequested = false and isFullWidthAligned_ = true by writing some bug in the code.

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js [hunk 11]
```diff
@@ -144,6 +169,36 @@ export class AmpAd3PImpl extends AMP.BaseElement {
...
+      return this.attemptFullWidthSizeChange_();
+    }
+  }
+
+  /**
+   * Sets the fullWidthState_ variable.
+   * @private
+   */
+  computeFullWidthState_() {
+    const hasFullWidth = this.element.hasAttribute('data-full-width');
```
[the remark this answers is not in the document]
Comment (riklund) [reply extensions/amp-ad/0.1/amp-ad-3p-impl.js]: Done.

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js (hunk 8 above)
Comment (riklund) [reply extensions/amp-ad/0.1/amp-ad-3p-impl.js]: Done.

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js [hunk 12]
```diff
@@ -144,6 +169,36 @@ export class AmpAd3PImpl extends AMP.BaseElement {
...
+    const hasFullVW = this.element.getAttribute('width') == '100vw';
+    const fullWidthRatio = this.config.fullWidthHeightRatio;
+    const okFullWidthRatio =
+        !!fullWidthRatio && fullWidthRatio > 0.2 && fullWidthRatio < 5;
```
[the remark this answers is not in the document]
Comment (riklund) [reply extensions/amp-ad/0.1/amp-ad-3p-impl.js]: Done.

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js [hunk 13]
```diff
@@ -300,4 +375,32 @@ export class AmpAd3PImpl extends AMP.BaseElement {
...
+
+  /**
+  * Calculates the appropriate height & width for a responsive full width ad
+  * unit.
+  * @return {!Promise}
+  * @private
+  */
+  attemptFullWidthSizeChange_() {
+    const viewportSize = this.getViewport().getSize();
+    const minHeight = 100;
```
[the remark this answers is not in the document]
Comment (riklund) [reply extensions/amp-ad/0.1/amp-ad-3p-impl.js]: What's the advantage of moving it outside the class? The style guide suggests local variables should be declared close to where they are needed to minimize their scope (https://google.github.io/styleguide/jsguide.html#features-declared-when-needed).

100 and 500 are a pretty arbitrary choice of some bounds that seemed reasonable for checking the sanity of the height. I'm open for discussion.

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js [hunk 14]
```diff
@@ -300,4 +375,32 @@ export class AmpAd3PImpl extends AMP.BaseElement {
...
+    const maxHeight = Math.min(500, viewportSize.height);
```
[the remark this answers is not in the document]
Comment (riklund) [reply extensions/amp-ad/0.1/amp-ad-3p-impl.js]: Ditto - what's the advantage of this?

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js [hunk 15]
```diff
@@ -300,4 +375,32 @@ export class AmpAd3PImpl extends AMP.BaseElement {
...
+    const ratio = this.config.fullWidthHeightRatio;
+    dev().assert(!!ratio);
```
[the remark this answers is not in the document]
Comment (riklund) [reply extensions/amp-ad/0.1/amp-ad-3p-impl.js]: Done.

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js [hunk 16]
```diff
@@ -357,4 +359,170 @@ describes.realWin('amp-ad-3p-impl', {
       });
     });
   });
+
+  describe('responsive', () => {
```
[the remark this answers is not in the document]
Comment (riklund) [reply extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: Done.

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js [hunk 17]
```diff
@@ -357,4 +359,170 @@ describes.realWin('amp-ad-3p-impl', {
...
+
+    const VIEWPORT_WIDTH = 300;
+    const VIEWPORT_HEIGHT = 600;
+
+    beforeEach(() => {
+      adConfig['_ping_'].fullWidthHeightRatio = 1.2;
+      win.document.body.removeChild(ad3p.element);
+    });
+
+    describe('#buildCallback', () => {
```
[the remark this answers is not in the document]
Comment (riklund) [reply extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: Done.

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js [hunk 18]
```diff
@@ -357,4 +359,170 @@ describes.realWin('amp-ad-3p-impl', {
...
+      let iframe;
+      let impl;
+      let element;
+
+      function constructImpl(config) {
```
[the remark this answers is not in the document]
Comment (riklund) [reply extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: This function is quite different from createAmpAd so I don't really see how I could reconcile them, but open to suggestions.

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js [hunk 19]
```diff
@@ -357,4 +359,170 @@ describes.realWin('amp-ad-3p-impl', {
...
+        iframe = doc.createElement('iframe');
+
+        config.type = '_ping_';
+        element = createElementWithAttributes(doc, 'amp-ad', config);
+        element.appendChild(iframe);
```
[the remark this answers is not in the document]
Comment (riklund) [reply extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: I think an ad unit normally contains an iframe so this keeps it close to reality (I tried to mimic https://github.com/ampproject/amphtml/pull/11231/files here).

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js [hunk 20]
```diff
@@ -357,4 +359,170 @@ describes.realWin('amp-ad-3p-impl', {
...
+        impl.getViewport().getSize =
+          () => ({width: VIEWPORT_WIDTH, height: VIEWPORT_HEIGHT});
+      }
+
+      it('should do nothing for non-responsive', () => {
+        constructImpl({
+          width: '280',
+          height: '280',
+        });
+        expect(impl.buildCallback()).to.be.undefined;
```
[the remark this answers is not in the document]
Comment (riklund) [reply extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: Done.

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js [hunk 21]
```diff
@@ -357,4 +359,170 @@ describes.realWin('amp-ad-3p-impl', {
...
+          width: '100vw',
+          height: '280',
+          'data-auto-format': 'rspv',
+          'data-full-width': '',
+        });
+
+        const callback = impl.buildCallback();
+        expect(callback).to.exist;
+
+         // The returned promise fails for some reason.
```
[the remark this answers is not in the document]
Comment (riklund) [reply extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: Done.

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js [hunk 22]
```diff
@@ -357,4 +359,170 @@ describes.realWin('amp-ad-3p-impl', {
...
+        return callback.then(() => {
+          expect(element.offsetWidth).to.equal(VIEWPORT_WIDTH);
+          expect(element.offsetHeight).to.equal(250);
+        });
+      });
+    });
```
[the remark this answers is not in the document]
Comment (riklund) [reply extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: Done.

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js [hunk 23]
```diff
@@ -357,4 +359,170 @@ describes.realWin('amp-ad-3p-impl', {
...
+    describe('#onLayoutMeasure', () => {
```
[the remark this answers is not in the document]
Comment (riklund) [reply extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: Done.

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js (hunk 5 above)
Comment (riklund) [reply extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: As above - I don't see how, but open to suggestions.

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js (hunk 3 above)
Comment (riklund) [reply extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: Done.

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js (hunk 7 above)
Comment (riklund) [reply extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: Done.

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js (hunk 4 above)
Comment (riklund) [reply extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: Done.

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js (hunk 6 above)
Comment (riklund) [reply extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: Done.

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js (hunk 1 above)
Comment (riklund) [reply extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: Done.

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js (hunk 2 above)
Comment (riklund) [reply extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: Done.

**Comment by riklund** [comment]
Addressed comments.

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js [hunk 24]
```diff
@@ -144,6 +169,36 @@ export class AmpAd3PImpl extends AMP.BaseElement {
...
+   * Sets the fullWidthState_ variable.
+   * @private
+   */
+  computeFullWidthState_() {
+    const hasFullWidth = this.element.hasAttribute('data-full-width');
+    if (!hasFullWidth) {
+      this.fullWidthState_ = FullWidthState.NOT_ENABLED;
+      return;
+    }
+    user().assert(this.element.getAttribute('width') == '100vw',
```
Comment (lannka) [inline extensions/amp-ad/0.1/amp-ad-3p-impl.js]: shall we just override width to 100vw if we see `data-full-width`?

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (lannka) [reply extensions/amp-ad/0.1/amp-ad-3p-impl.js]: I agree with @zhouyx . We don't even need the `isFullWidthAligned_` boolean, since I don't see your code checks its value.

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (lannka) [reply extensions/amp-ad/0.1/amp-ad-3p-impl.js]: This is not a local variable, but a constant. For constant, AMP follows the convention of CAPTIALIZED naming.

If the number is arbitrary that you just made up for AMP, I suggest we can just remove them.

File: examples/a4a-fullwidth.amp.html (hunk 9 above)
[the remark this answers is not in the document]
Comment (lannka) [reply examples/a4a-fullwidth.amp.html]: did you have a working example for adsense?

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js [hunk 25]
```diff
@@ -357,4 +357,171 @@ describes.realWin('amp-ad-3p-impl', {
...
+      let iframe;
+      let impl;
+      let element;
+      let attemptChangeSizeSpy;
+
+      function constructImpl(config) {
+        iframe = win.document.createElement('iframe');
+
+        config.type = '_ping_';
+        element = createElementWithAttributes(win.document, 'amp-ad', config);
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: seems like some code duplication with `createAmpAd` function at the top of this file.

File: examples/a4a-fullwidth.amp.html (hunk 9 above)
[the remark this answers is not in the document]
Comment (riklund) [reply examples/a4a-fullwidth.amp.html]: Yes, this file is a working example. It seems to use the 3p path locally (for me at least) meaning full width doesn't work before but works after this change.

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js (hunk 24 above)
Comment (riklund) [reply extensions/amp-ad/0.1/amp-ad-3p-impl.js]: Are we allowed to always do that successfully at this point?
If not let's not do it - from a product perspective I think it's better for us to consistently fail than to work in some cases.

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js [hunk 26]
```diff
@@ -144,6 +169,36 @@ export class AmpAd3PImpl extends AMP.BaseElement {
...
+        'Ad units with data-full-width must have width="100vw".');
+    const fullWidthRatio = this.config.fullWidthHeightRatio;
+
+    // Somewhat arbitrary bounds for ad aspect ratios to ensure they are
+    // reasonably sane.
+    const lowerBound = 0.2;
+    const upperBound = 5;
+    user().assert(fullWidthRatio >= lowerBound && fullWidthRatio <= upperBound,
+        'Ad network does not support full width ads.');
```
[the remark this answers is not in the document]
Comment (riklund) [reply extensions/amp-ad/0.1/amp-ad-3p-impl.js]: The idea here was to have some sanity checks of the actual value in the network config.

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js (hunk 25 above)
Comment (riklund) [reply extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: Ack, (also see https://github.com/ampproject/amphtml/pull/12200#discussion_r153683396). 
I don't really see how we can share much code here in a meaningful way but open to suggestions.

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (riklund) [reply extensions/amp-ad/0.1/amp-ad-3p-impl.js]: Done.

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js (hunk 26 above)
[the remark this answers is not in the document]
Comment (lannka) [reply extensions/amp-ad/0.1/amp-ad-3p-impl.js]: the sanity check should have been done when the constant enters in the code base. Every error message got full bytes in our precious binary size.

[the remark this answers is not in the document]
Comment (riklund) [reply extensions/amp-ad/0.1/amp-ad-3p-impl.js]: Done.

**Comment by riklund** [comment]
Thanks!

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js [hunk 27]
```diff
@@ -300,4 +367,32 @@ export class AmpAd3PImpl extends AMP.BaseElement {
...
+  * Calculates and attempts to set the appropriate height & width for a
+  * responsive full width ad unit.
+  * @return {!Promise}
+  * @private
+  */
+  attemptFullWidthSizeChange_() {
+    const viewportSize = this.getViewport().getSize();
+    const maxHeight = Math.min(MAX_FULL_WIDTH_HEIGHT, viewportSize.height);
+    const ratio = this.config.fullWidthHeightRatio;
+    dev().assert(!!ratio,
```
Comment (lannka) [inline extensions/amp-ad/0.1/amp-ad-3p-impl.js]: this is already checked in the `shouldRequestFullWidth_` function

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js [hunk 28]
```diff
@@ -144,6 +168,29 @@ export class AmpAd3PImpl extends AMP.BaseElement {
...
+  shouldRequestFullWidth_() {
+    const hasFullWidth = this.element.hasAttribute('data-full-width');
+    if (!hasFullWidth) {
+      return false;
+    }
+    user().assert(this.element.getAttribute('width') == '100vw',
+        'Ad units with data-full-width must have width="100vw".');
+    user().assert(!!this.config.fullWidthHeightRatio,
+        'Ad network does not support full width ads.');
+    dev().info(TAG_3P_IMPL, 'Full width requested');
```
Comment (lannka) [inline extensions/amp-ad/0.1/amp-ad-3p-impl.js]: can you attach the resource ID to the log:

`#${this.getResource().getId()} Full width requested`

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js (hunk 25 above)
Comment (lannka) [reply extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: for instance, you should not create an iframe and append it to the amp-ad element manually. and i'm not seeing you read this iframe anyway.

File: ads/_config.js [hunk 29]
```diff
@@ -48,6 +49,11 @@ let AdNetworkConfigDef;
  *   // We highly recommend all networks to implement the API,
  *   // see details in the README.md
  *   renderStartImplemented: boolean
+ *
+ *   // The width / height ratio for full width ad units.
+ *   // If absent, it means the network does not support full width ad units.
+ *   // Example value: 1.2
+ *   fullWidthHeightRatio: number
```
Comment (lannka) [inline ads/_config.js]: if you want to force a range for this number, i suggest you add it to the documentation of this feature (which can be a separate PR).

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js (hunk 27 above)
Comment (riklund) [reply extensions/amp-ad/0.1/amp-ad-3p-impl.js]: Yes. 
That's why it's a dev assert and noted as "should never happen" - it's here as help for future developers.

File: ads/_config.js (hunk 29 above)
Comment (riklund) [reply ads/_config.js]: Sounds good - haven't included it in this PR.

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js (hunk 28 above)
Comment (riklund) [reply extensions/amp-ad/0.1/amp-ad-3p-impl.js]: Done.

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js (hunk 25 above)
Comment (riklund) [reply extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: Done.

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js (hunk 27 above)
Comment (lannka) [reply extensions/amp-ad/0.1/amp-ad-3p-impl.js]: I don't see this duplicated check needed. again, all error message takes real bytes in the binary.

**Comment by riklund** [comment]
Done, thanks!

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js [hunk 30]
```diff
@@ -357,4 +357,165 @@ describes.realWin('amp-ad-3p-impl', {
...
+    describe('should resize correctly', () => {
+      let impl;
+      let element;
+      let attemptChangeSizeSpy;
+
+      function constructImpl(config) {
+
+        config.type = '_ping_';
+        element = createElementWithAttributes(win.document, 'amp-ad', config);
+        document.body.appendChild(element);
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: This might be the reason why the tests are failing. You're accessing global object which will affect all other tests.
